tests/cpp/test_multidevice_lower_communication.cpp

```diff
@@ -202,6 +202,38 @@ TEST_F(LowerCollectiveTest, Allgather) {
   EXPECT_TRUE(at::equal(out_tensor, unsharded_tensor));
 }
 
+TEST_F(LowerCollectiveTest, Allgather_LoopSplit) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const auto num_devices = communicator_->size();
+  auto mesh = DeviceMesh::createForNumDevices(num_devices);
+
+  TensorView* in = makeContigTensor(1);
+  in->setDeviceMesh(mesh);
+  TensorView* out = set(in);
+  fusion->addInput(in);
+  fusion->addOutput(out);
+
+  in->split(0, num_devices, /*inner_split=*/false);
+  in->axis(0)->parallelize(ParallelType::DIDx);
+  in->setAllocationDomain(in->getLoopDomain(), true);
+
+  out->split(0, num_devices, /*inner_split=*/false);
+  out->setAllocationDomain(out->getLoopDomain(), true);
+
+  at::Tensor unsharded_tensor =
+      at::randn({num_devices * kTensorSize}, at::kFloat);
+  at::Tensor in_tensor =
+      shardTensor(unsharded_tensor, in).to(communicator_->device());
+
+  FusionExecutorCache fec(std::move(fusion));
+  at::Tensor out_tensor = fec.runFusionWithInputs({in_tensor})[0];
+  assertIsCompiledToHostIrContainer(fec);
+
+  EXPECT_TRUE(at::equal(out_tensor.cpu(), unsharded_tensor));
```

Comment:
Why not use `validate` here?

Reply:
I noticed allgather's lowering was not changed...I'm a bit surprised it didn't need any modifications for inputs with DID loop split! I might have missed a few earlier PRs though

Reply:
> Why not use `validate` here?

Since `validate` allows for (small) differences, if two tensors are supposed to be exactly the same, just using the simpler validation method, i.e., `at::equal`, would be more preferable.

Reply:
> I'm a bit surprised it didn't need any modifications for inputs with DID loop split!

Whether we call lowerToAllGather depends on I/O meshes and whether I/O is sharded: https://github.com/NVIDIA/Fuser/blob/67127c94a43e23f673cab05f769269f38e4fb043/csrc/multidevice/lower_communication.cpp#L285. isSharded have been reading the allocation domain ince #3444. 

That being said, I think this PR as is is a bit too permissive and may lower a `set` to `Allgather` without properly checking its allocation domain. For example, https://github.com/NVIDIA/Fuser/blob/67127c94a43e23f673cab05f769269f38e4fb043/csrc/multidevice/utils.cpp#L77 reads root and logical and needs to be updated. I'll try to fix that.

Reply:
> That being said, I think this PR as is is a bit too permissive and may lower a set to Allgather without properly checking its allocation domain.

I tried to address this in https://github.com/NVIDIA/Fuser/pull/3284#discussion_r1870826379.